packages/block/block.ts

```diff
@@ -0,0 +1,220 @@
+/* eslint-disable @typescript-eslint/unbound-method */
+import {
+  childNodes$,
+  cloneNode$,
+  createEventListener,
+  insertBefore$,
+  insertText,
+  innerHTML$,
+  remove$ as removeElement$,
+  setAttribute,
+  setText,
+  mapGet$,
+  mapHas$,
+  mapSet$,
+} from './dom';
+import { renderToTemplate } from './template';
+import { AbstractBlock, EditType, Hole } from './types';
+import type { Edit, EditChild, Props, VElement } from './types';
+
+const holeCache = new Map();
+export const HOLE_PROXY = new Proxy(
+  {},
+  {
+    // A universal getter will return a Hole instance if props[any] is accessed
+    // Allows code to identify holes in virtual nodes ("digs" them out)
+    get(_, prop: string): Hole {
+      if (mapHas$.call(holeCache, prop)) return mapGet$.call(holeCache, prop);
+      const hole = new Hole(prop);
+      mapSet$.call(holeCache, prop, hole);
+      holeCache.set(prop, hole);
```

Comment:
Is it possible to leak memory here? It is possible to have infinite props, so the Map size will grow indefinitely.

Reply:
good point, but it is also good to have a cache here (instantiating a lot of holes can be wasteful), any ideas for a good middle ground?

Reply:
> any ideas for a good middle ground?

I have some rough ideas, but they all have drawbacks:

- limit size cache or even a limit size LRU cache to prevent memory leak
- the memo (deep equal -> referential equality) to prevent infinite props

Reply:
I just came up with an example:

```ts
const block = createBlock(/** */);
for (let i = 0; i < 1000000; i++) {
  block({ title: 'Hello World' });
}
```

Here is the problem, although the value is always the same, the referential equality doesn't retain as the props are created as an object literal (`Object.is({ title: 'Hello World' }, { title: 'Hello World' }) === false`). And it will cause a memory leak.

Reply:
I have an idea. We should not have a single `HOLE_PROXY` across the entire library.

We should create a `hole proxy` instance for every block when `createBlock` is invoked. So every block created will have a `hole proxy` instance bound to them.

And instead of having an infinite cache, a block only has to remember the last prop value (a cache with size 1). Only when the prop changes, the block will re-create a `hole proxy` instance.

Reply:
Got it! Let me make that change